### PR TITLE
GH-166: Fix `ChannelBinder` config inconsistency

### PR DESCRIPTION
**Pull request by artembilan**
Fixes GH-166 (https://github.com/spring-cloud/spring-cloud-stream/issues/166)

Since `onInit()` in the `MessageChannelBinderSupport` was called before the `evaluationContext` population,
the `RedisMessageChannelBinder#errorAdapter` caused `NPE` for its `evaluationContext`.

* Change the initialization order
* Use `ExpressionUtils` to populate `evaluationContext` - creates `StandardEvaluationContext` if there is no one in the `applicationContext`
* Add more initializations for the `RedisMessageChannelBinder#errorAdapter`
* Remove explicit `evaluationContext` population from the `AbstractTestBinder` to rely on the fix for this ticket
* Some other simple polishing

### CODE REVIEW

File: spring-cloud-stream-binders/spring-cloud-stream-binder-spi/src/main/java/org/springframework/cloud/stream/binder/MessageChannelBinderSupport.java [hunk 1]
```diff
@@ -602,18 +603,17 @@ protected final MessageValues deserializePayloadIfNecessary(Message<?> message)
 	}
 
 	protected final MessageValues deserializePayloadIfNecessary(MessageValues message) {
-		MessageValues messageToSend = message;
 		Object originalPayload = message.getPayload();
-		MimeType contentType = this.contentTypeResolver.resolve(messageToSend);
+		MimeType contentType = this.contentTypeResolver.resolve(message);
 		Object payload = deserializePayload(originalPayload, contentType);
```
Comment (garyrussell) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-spi/src/main/java/org/springframework/cloud/stream/binder/MessageChannelBinderSupport.java]: Why this change? It is considered bad practice (and Sonar complains about it) to modify a parameter value.

Comment (artembilan) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-spi/src/main/java/org/springframework/cloud/stream/binder/MessageChannelBinderSupport.java]: `MessageValues messageToSend = message;` - are two vars for the same object, so any modifications have effect

File: spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/main/java/org/springframework/cloud/stream/binder/redis/RedisMessageChannelBinder.java [hunk 2]
```diff
@@ -165,6 +165,8 @@ public RedisMessageChannelBinder(RedisConnectionFactory connectionFactory,
 	@Override
 	protected void onInit() {
 		this.errorAdapter.setIntegrationEvaluationContext(this.evaluationContext);
+		this.errorAdapter.setBeanFactory(getBeanFactory());
+		this.errorAdapter.afterPropertiesSet();
```
Comment (mbogoevici) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/main/java/org/springframework/cloud/stream/binder/redis/RedisMessageChannelBinder.java]: Looking at how the `errorAdapter` is setup, I think we'd best move its entire initialization in `onInit()` - makes it an easier read than the current split.

File: spring-cloud-stream-binders/spring-cloud-stream-binder-spi/src/main/java/org/springframework/cloud/stream/binder/MessageChannelBinderSupport.java [hunk 3]
```diff
@@ -377,10 +378,10 @@ public void setDefaultDurableSubscription(boolean defaultDurableSubscription) {
 	@Override
 	public void afterPropertiesSet() throws Exception {
 		Assert.notNull(this.applicationContext, "The 'applicationContext' property cannot be null");
-		onInit();
 		if (this.evaluationContext == null) {
-			this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+			this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
 		}
+		onInit();
```
Comment (mbogoevici) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-spi/src/main/java/org/springframework/cloud/stream/binder/MessageChannelBinderSupport.java]: With this move, I think we could remove `onInit()` entirely and just rely on `super.afterPropertiesSet()` calls in subclasses.

Comment (artembilan) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-spi/src/main/java/org/springframework/cloud/stream/binder/MessageChannelBinderSupport.java]: Yes, makes sense, too.
Having all this `MessageChannelBinderSupport` hierarchy as an internal stuff, there is no so much worry about end-users who won't call `super.afterPropertiesSet()` from his impl.

File: spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/main/java/org/springframework/cloud/stream/binder/redis/RedisMessageChannelBinder.java (hunk 2 above)
Comment (artembilan) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/main/java/org/springframework/cloud/stream/binder/redis/RedisMessageChannelBinder.java]: I don't mind. Will do together with the `afterPropertiesSet()` refactoring.

**Comment by artembilan** [comment]
Pushed, but I left `errorAdapter` instantiation in the ctor to keep it as `final` for better performance.

**Comment by mbogoevici** [comment]
OK, we'll just need to remember not to touch any of this. 

LGTM, will merge.

**Comment by mbogoevici** [comment]
Rebased, squashed, merged https://github.com/spring-cloud/spring-cloud-stream/commit/2e7f9cc729d37651985eff5485bb684f34c1e2a0
